sklearn/neighbors/_binary_tree.pxi

```diff
@@ -1053,9 +1053,13 @@ cdef class BinaryTree:
         if leaf_size < 1:
             raise ValueError("leaf_size must be greater than or equal to 1")
 
+        longest_data = max(len(item) for item in data)
+        if np.any([len(item) < longest_data for item in data]):
+            raise ValueError("Input points must have "
+                             "the same number of dimensions")
+
         n_samples = data.shape[0]
         n_features = data.shape[1]
-
         self.data_arr = np.asarray(data, dtype=DTYPE, order='C')
```

Comment:
I think you should rather just use `sklearn.utils.check_array`:

```python
        self.data_arr = check_array(data, dtype=DTYPE, order='C')
```

It should display a `ValueError` with the message `setting an array element with a sequence.` which is good enough to me.